url9ecf1e8297_nhamidi_ninja_tgz-pJ8-net_fec_openrq_test_TestRunnerJ8/src/RQLibrary/Encoder.java

```diff
@@ -0,0 +1,2552 @@
+		
+		// increment the total number of source symbols
+		num_source_symbols++;
+	    } else // is a repair symbol
+	    {
+		// add to the set of repair symbols
+		repair_symbols.add(enc_symbol);
+		
+		// increment the total number of repair symbols
+		num_repair_symbols++;
+	    }
+	}
+	
+	
+	
+	/*
+	 * print topology of the received encoding symbols
+	 */
+	StringBuilder st = new StringBuilder();
+	st.append("Symbols topology: \n");
+	st.append("# Source: ");
+	st.append(num_source_symbols);
+	st.append("\n# Repair: ");
+	st.append(num_repair_symbols);
+	st.append("\n Repair indexes:\n");
+	for (EncodingSymbol i : repair_symbols)
+	    st.append(i.getESI() + ", ");
+	st.append("\n# Missing: ");
+	st.append(num_missing_symbols);
+	st.append("\n Missing indexes:\n");
+	for (Integer i : missing_symbols)
+	    st.append(i + ", ");
+	// System.out.println(st.toString());
+	/*
+	 * end print
+	 */
+	
+	// are there enough repair symbols to compensate the missing source symbols?
+	if ( num_repair_symbols < num_missing_symbols ) {
+	    throw new RuntimeException(Integer.toString(num_missing_symbols - num_repair_symbols)); // TODO shouldn't be runtime exception (too generic)
+	    // throw new RuntimeException( Integer.toString(num_missing_symbols)); // TODO shouldn't be runtime exception (too generic)
+	}
+	
+	// number of extra repair symbols to be used for the decoding process
+	int overhead = num_repair_symbols - missing_symbols.size();
+	
+	// number of rows in the decoding matrix
+	int M = L + overhead;
+	
+	// maps the ESI of a symbol to its line in the decoding matrix
+	Map<Integer, byte[]> esiToLTCode = new TreeMap<Integer, byte[]>();
+	
+	// allocate buffer for the decoded data
+	byte[] decoded_data = new byte[K * T];
+	
+	// all source symbols received! :D
+	if ( num_source_symbols == K ) {
+	    // collect all payloads from the source symbols
+	    for (EncodingSymbol enc_symbol : source_symbols) {
+		System.arraycopy(enc_symbol.getData(), 0, decoded_data, enc_symbol.getESI() * T, T);
+	    }
+	    
+	    // return the "decoded" source block
+	    return new SourceBlock(eb.getSBN(), decoded_data, T, K);
+	}
+	
+	// not so lucky...
+	else {
+	    
+	    // allocate memory for the decoding matrix
+	    byte[][] constraint_matrix = new byte[M][];
+	    
+	    // generate the original constraint matrix
+	    byte[][] lConstraint = generateConstraintMatrix(kLinha, T);
+	    
+	    // copy to our decoding matrix
+	    for (int row = 0; row < L; row++)
+		constraint_matrix[row] = lConstraint[row];
+	    
+	    // initialize D
+	    byte[][] D = new byte[M][T];
+	    
+	    // populate D with the received source symbols
+	    if ( num_source_symbols != 0 ) {
+		
+		Iterator<EncodingSymbol> it = source_symbols.iterator();
+		
+		do {
+		    EncodingSymbol source_symbol = (EncodingSymbol) it.next();
+		    
+		    D[source_symbol.getESI() + S + H] = source_symbol.getData();
+		    
+		} while (it.hasNext());
+	    }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 517:527, 711:720.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.